core/server/worker/src/main/java/alluxio/worker/block/TieredBlockStore.java

```diff
@@ -526,42 +697,92 @@ private void abortBlockInternal(long sessionId, long blockId) throws BlockDoesNo
     try {
-      // When committing TempBlockMeta, the final BlockMeta calculates the block size according to
-      // the actual file size of this TempBlockMeta. Therefore, commitTempBlockMeta must happen
-      // after moving actual block file to its committed path.
-      BlockStoreLocation loc;
-      String srcPath;
-      String dstPath;
-      TempBlockMeta tempBlockMeta;
-      try (LockResource r = new LockResource(mMetadataReadLock)) {
-        checkTempBlockOwnedBySession(sessionId, blockId);
-        tempBlockMeta = mMetaManager.getTempBlockMeta(blockId);
-        srcPath = tempBlockMeta.getPath();
-        dstPath = tempBlockMeta.getCommitPath();
-        loc = tempBlockMeta.getBlockLocation();
+      // Allocate from given location.
+      dirView = mAllocator.allocateBlockWithView(sessionId, options.getSize(),
+          options.getLocation(), allocatorView);
+
+      if (dirView != null) {
+        return dirView;
       }
 
-      // Heavy IO is guarded by block lock but not metadata lock. This may throw IOException.
-      FileUtils.move(srcPath, dstPath);
+      if (options.isForceLocation()) {
+        if (options.isEvictionAllowed()) {
+          freeSpace(sessionId, options.getSize(), options.getSize(), options.getLocation());
+          dirView = mAllocator.allocateBlockWithView(sessionId, options.getSize(),
+              options.getLocation(), allocatorView.refreshView());
 
-      try (LockResource r = new LockResource(mMetadataWriteLock)) {
-        mMetaManager.commitTempBlockMeta(tempBlockMeta);
-      } catch (BlockAlreadyExistsException | BlockDoesNotExistException
-          | WorkerOutOfSpaceException e) {
-        throw Throwables.propagate(e); // we shall never reach here
-      }
+          if (dirView == null) {
+            LOG.error("Target tier: {} has no evictable space to store {} bytes for session: {}",
+                options.getLocation(), options.getSize(), sessionId);
+            return null;
+          }
+        } else {
+          LOG.error("Target tier: {} has no available space to store {} bytes for session: {}",
+              options.getLocation(), options.getSize(), sessionId);
+          return null;
+        }
+      } else {
+        dirView = mAllocator.allocateBlockWithView(sessionId, options.getSize(),
+            BlockStoreLocation.anyTier(), allocatorView);
 
-      // Check if block is pinned on commit
-      if (pinOnCreate) {
-        addToPinnedInodes(BlockId.getFileId(blockId));
-      }
+        if (dirView != null) {
+          return dirView;
+        }
 
-      return loc;
-    } finally {
-      mLockManager.unlockBlock(lockId);
+        if (options.isEvictionAllowed()) {
+          // There is no space left on worker.
+          // Free more than requested by configured free-ahead size.
+          long freeAheadBytes =
+              ServerConfiguration.getBytes(PropertyKey.WORKER_TIERED_STORE_FREE_AHEAD_BYTES);
```

Comment:
Is this concept necessary, it seems a bit too much of a heuristic?

Reply:
This concept is a fail-safe against a corner case:
- Worker reached peak capacity
- Annotator is LRFU
- Concurrent readers/writers

In this case, each allocation could end up sorting the whole block list in order to find eviction candidate. This is due to LRFU being an online scheme. If CPU starts burning, we could evict more than necessary using this prop and avoid excessive sorting per allocation.

This should be mentioned in the tuning guides.